ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixClusterManager.java

```diff
@@ -1153,13 +1154,31 @@ public List<AmbryReplica> getReplicaIdsByState(AmbryPartition partition, Replica
         RoutingTableSnapshot routingTableSnapshot =
             clusterMapConfig.clusterMapUseAggregatedView ? globalRoutingTableSnapshotRef.get()
                 : dcToRoutingTableSnapshotRef.get(dc).get();
+        Collection<String> resourcesFromRoutingTable = routingTableSnapshot.getResources();
         routingTableSnapshot.getInstancesForResource(resourceName, partition.toPathString(), state.name())
             .stream()
             .map(instanceConfig -> instanceNameToAmbryDataNode.get(instanceConfig.getInstanceName()))
             .filter(dataNode -> dataNode.getDatacenterName().equals(dc))
             .map(dataNode -> ambryDataNodeToAmbryReplicas.get(dataNode).get(partition.toPathString()))
             .filter(Objects::nonNull)
             .forEach(replicas::add);
+        if (replicas.isEmpty()) {
+          // There is some change made to resources in IDEALSTATES and it's not yet reflected in EXTERNAALVIEW.
+          // Try to search the partition in all resources
+          for (String resource : resourcesFromRoutingTable) {
+            routingTableSnapshot.getInstancesForResource(resource, partition.toPathString(), state.name())
+                .stream()
+                .map(instanceConfig -> instanceNameToAmbryDataNode.get(instanceConfig.getInstanceName()))
+                .filter(dataNode -> dataNode.getDatacenterName().equals(dc))
+                .map(dataNode -> ambryDataNodeToAmbryReplicas.get(dataNode).get(partition.toPathString()))
+                .filter(Objects::nonNull)
+                .forEach(replicas::add);
+            if (!replicas.isEmpty()) {
+              // We found the partition
+              break;
```

Comment:
If we break out of for loop, we may not collect replicas in other dcs?

Reply:
it will break out the for loop to search resources, not the dcs.